File: Lecture037 Recursion Day7/subsets.cpp
class Solution {
private:
    void solve(vector<int> nums, vector<int> output, int index, vector<vector<int> >& ans) {
        //base case
        if(index >= nums.size()) {
            ans.push_back(output);
            return ;
        }
        
        //exclude
        solve(nums, output, index+1, ans);
        
        //include
        int element = nums[index];
        output.push_back(element);
        solve(nums, output, index+1, ans);
        
    }
    
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        
        vector<vector<int> > ans;
        vector<int> output;
        int index = 0;
        solve(nums, output, index, ans);
        return ans;
        
    }
};














// Faster approach
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> v;
        int i, j, k;
        for(i=0; i<(1<<n); i++){
            vector<int> temp;
            for(j=0; j<n; j++){
                if(i&(1<<j)){
                    temp.push_back(nums[j]);
                }
            }
            v.push_back(temp);
        }
        return v;
    }
};

File: Lecture037 Recursion Day7/subsets.java
class Solution {
	private void solve(int[] a,int index,List<Integer> output,List<List<Integer>> ans){
		if(index >= a.length){
			ans.add(new ArrayList<>(output));
			return;
		}
		solve(a,index+1,new ArrayList<>(output),ans);
		int ele = a[index];
		output.add(ele);
		solve(a,index+1,output,ans);
		
	}
	
	List<List<Integer>> subsets(int[] A) {
	    // add your logic here
		List<List<Integer>> ans = new ArrayList<>();
		List<Integer> output = new ArrayList<>();
		int index = 0;
		solve(A,index,output,ans);
		return ans;
	}
}
